### ISSUE TITLE
Cannot Create FargateTaskDefinition with ECR image

### ISSUE THREAD

**Issue by realharry** [aws/aws-cdk#3644]
* **I'm submitting a ...**
  - [x] :beetle: bug report
  - [ ] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/aws/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*

1. Create a VPC
1. Create an application ELB, targetGroup, and applicationListener.
1. Create a fargate cluster.
1. Try to create a `FargateTaskDefinition` having a container with an ECR image. Specify `ExecutionRole` with at least `AmazonECSTaskExecutionRolePolicy` and `AmazonEC2ContainerRegistryFullAccess`.

It throws the following error upon `cdk deploy`:

```BASH
 0/8 | 4:14:06 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata                        | CDKMetadata
 0/8 | 4:14:06 PM | CREATE_IN_PROGRESS   | AWS::ElasticLoadBalancingV2::TargetGroup  | website-tg (websitetgE0485AEE)
 0/8 | 4:14:06 PM | CREATE_IN_PROGRESS   | AWS::ECS::Cluster                         | website-cluster (websiteclusterE145DAC4)
 0/8 | 4:14:06 PM | CREATE_IN_PROGRESS   | AWS::ECS::TaskDefinition                  | website-task (websitetask43225E36)
 0/8 | 4:14:06 PM | CREATE_IN_PROGRESS   | AWS::EC2::SecurityGroup                   | website-sg (websitesg49C6853E)
 1/8 | 4:14:06 PM | CREATE_FAILED        | AWS::ECS::TaskDefinition                  | website-task (websitetask43225E36) Fargate requires task definition to have execution role ARN to support ECR images. (Service: AmazonECS; Status Code: 400; Error Code: ClientException; Request ID: c5bb139d-f68a-47c3-9c32-a096508313bf)
	new TaskDefinition (/Users/harry/Projects/devops/deploy/rbi5-main/node_modules/@aws-cdk/aws-ecs/lib/base/task-definition.ts:267:21)
	\_ new FargateTaskDefinition (/Users/harry/Projects/devops/deploy/rbi5-main/node_modules/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts:78:5)
	\_ Object.createTaskDefinition (/Users/harry/Projects/devops/deploy/rbi5-main/lib/rbi5-ecs-util.ts:20:16)
	\_ Rbi5MainStack.createClusterAndService (/Users/harry/Projects/devops/deploy/rbi5-main/lib/rbi5-main-stack.ts:90:28)
	\_ Rbi5MainStack.init (/Users/harry/Projects/devops/deploy/rbi5-main/lib/rbi5-main-stack.ts:75:33)
	\_ new Rbi5MainStack (/Users/harry/Projects/devops/deploy/rbi5-main/lib/rbi5-main-stack.ts:68:10)
	\_ Object.<anonymous> (/Users/harry/Projects/devops/deploy/rbi5-main/bin/rbi5-main.ts:18:1)
	\_ Module._compile (internal/modules/cjs/loader.js:701:30)
	\_ Module.m._compile (/Users/harry/Projects/devops/deploy/rbi5-main/node_modules/ts-node/src/index.ts:473:23)
	\_ Module._extensions..js (internal/modules/cjs/loader.js:712:10)
	\_ Object.require.extensions.(anonymous function) [as .ts] (/Users/harry/Projects/devops/deploy/rbi5-main/node_modules/ts-node/src/index.ts:476:12)
	\_ Module.load (internal/modules/cjs/loader.js:600:32)
	\_ tryModuleLoad (internal/modules/cjs/loader.js:539:12)
	\_ Function.Module._load (internal/modules/cjs/loader.js:531:3)
	\_ Function.Module.runMain (internal/modules/cjs/loader.js:754:12)
	\_ Object.<anonymous> (/Users/harry/Projects/devops/deploy/rbi5-main/node_modules/ts-node/src/bin.ts:158:12)
	\_ Module._compile (internal/modules/cjs/loader.js:701:30)
	\_ Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
	\_ Module.load (internal/modules/cjs/loader.js:600:32)
	\_ tryModuleLoad (internal/modules/cjs/loader.js:539:12)
	\_ Function.Module._load (internal/modules/cjs/loader.js:531:3)
	\_ Function.Module.runMain (internal/modules/cjs/loader.js:754:12)
	\_ findNodeScript.then.existing (/usr/local/lib/node_modules/npm/node_modules/libnpx/index.js:268:14)
```
 
* **What is the expected behavior (or behavior of feature suggested)?**

It should create a fargate task definition stack without errors.

* **What is the motivation / use case for changing the behavior or adding this feature?**

Docs, https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.FargateTaskDefinition.html , states that it should work when we provide proper executionRole name. I tried both names and ARNs for both executionRole and taskRole (what's the difference?) for all different combinations, but none worked.

* **Please tell us about your environment:**

  - CDK CLI Version: 1.3.0 (build bba9914)
  - Module Version: 1.3.0
  - OS: OSX Mojave
  - Language: TypeScript
  - Node: v10.15.3
  - Typescript: Version 3.5.3

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

**Comment by fitzee** [comment]
What does your container definition look like?

**Comment by realharry** [comment]
Hi, here's relevant code:

```Javascript
const createTaskDefinition = (scope: cdk.Construct, id: string, props: any): ecs.IFargateTaskDefinition => {
  const taskId = id + '-task';
  const task = new ecs.FargateTaskDefinition(scope, taskId, {
    cpu: 1024,
    memoryLimitMiB: 4096,
    executionRole: props.executionRole,
  });

  let registryId = `xyz.us-west-2.amazonaws.com/${id}`;
  const containerId = id;
  const container = task.addContainer(containerId, {
    image: ecs.ContainerImage.fromRegistry(registryId),
    memoryLimitMiB: 2048,
  });
  return task;
};
```

**Comment by pkandasamy91** [comment]
@realharry you should use ecs.ContainerImage.fromEcrRepository instead of ecs.ContainerImage.fromRegistry

Something like
```ts

taskDefinition.addContainer("web", {
        image: ecs.ContainerImage.fromEcrRepository(new Repository(stack, "myECRImage")),
        memoryLimitMiB: 2048
      });
```

**Comment by fitzee** [comment]
Something seems wrong, if you use fromRegistry(), then registryId needs to be a fully qualified path to an image.  You seem to be substituting the construct id, which I am not sure is what you want.   I follow these guidelines:

* If the full image location within ECR is known, e.g.:
```node
imgLoc = '111345817488.dkr.ecr.us-west-2.amazonaws.com/aws-appmesh-envoy:v1.9.1.0-prod'
```
then I use:
```node
EcrImage.fromRegistry(imgLoc)
```
* If I have created the repository within my CDK app and built my own images, I use:
```node
EcrImage.fromEcrRepository(repository, tag)
```
Where repository is the name and tag is something like 'latest' or a specific version.

* If I just know the name of my image, e.g.:
```node
imgName = 'amazon/aws-xray-daemon'
```
then I use:
```node
ContainerImage.fromRegistry(imgName)
```

Hope that makes a bit of sense!

**Comment by realharry** [comment]
@fitzee sorry that my example code was unclear. the registryId is valid. I use "id" for service names like "website" or "order-service", and the constructed full path points (by design) to the correct registry in ECR.

@pkandasamy91 I tried both `fromRegistry()` and `fromEcrRepository()`, but it didn't make a difference.

I do realize that my code has some layers which may obscure the real cause of the problem here. Let me try to simplify the code and test again. Thanks!

**Comment by realharry** [comment]
I played around with the code some more, and as it turned out my executionRole was not properly set. I was able to proceed further without the "Fargate requires task definition to have execution role ARN to support ECR images" error. But, the deployment still failed with the same symptoms/errors as reported in https://github.com/aws/aws-cdk/issues/3646. Closing this ticket  in leu of #3646 . Thanks!
